osutil/udev/netlink/rawsockstop.go

```diff
@@ -0,0 +1,63 @@
+package netlink
+
+import (
+	"fmt"
+	"math/bits"
+	"os"
+	"syscall"
+)
+
+// RawSockStopper returns a pair of functions to manage stopping code
+// reading from a raw socket, readableOrStop blocks until
+// fd is readable or stop was called.
+// TODO: with go 1.11+ it should be possible to just switch to setting
+// fd to non-blocking and then wrapping the socket via os.NewFile and
+// use Close to force a read to stop.
+// c.f. https://github.com/golang/go/commit/ea5825b0b64e1a017a76eac0ad734e11ff557c8e
+func RawSockStopper(fd int) (readableOrStop func() (bool, error), stop func(), err error) {
+	stopR, stopW, err := os.Pipe()
```

Comment:
As mentioned yesterday it currently fails on nested hotplug tests with "udevmon.go:148: udev event error: Internal error: bad file descriptor" when processing normal events (what was confusing is that this error occured in random moments) - at some point the udevmon would not register an add or remove event.

The problem is with `os.File*`s returned by `os.Pipe()` going out of scope here and eventually getting garbage-collected, making the descriptors invalid. Here is one way of fixing it by wrapping them in a "stopper" object: https://paste.ubuntu.com/p/nkhPXFNRXP/
(it can probably be slightly simplified)

Reply:
You should use `runtime.KeepAlive` to ensure lifecycle is handled correctly. It is used for most raw file descriptor code.

Reply:
ok, the fix for that should be simple, close over stopR instead of reading out the Fd once